src/runtime/substitution.cc

```diff
@@ -1929,6 +1942,107 @@ void GraphSearchHelper::graph_optimize(
   optimal_views = real_optimal_views;
 }
 
+/**
+ * @brief Experimental DP algorithm to optimize PCG with the consideration of
+ * memory usage. This is to avoid polluting the current Unity search algorithm
+ * above. And this should be merged to GraphSearchHelper::graph_optimize
+ * eventually.
+ *
+ * @param[in] budget Not used
+ * @param[in] only_data_parallel Not used
+ * @param[out] best_graph The best possible PCG after optimization
+ * @param[out] optimal_views The corresponding device placement views of the
+ * best graph
+ * @param[out] search_result The performance result of the search
+ */
+void GraphSearchHelper::graph_optimize_with_memory(
```

Comment:
This function is rather long, is there any way to break it up?

Reply:
The previous implementation probably also should have been broken up :slightly_smiling_face:. A bunch of this code was written at 2am in the couple weeks leading up the OSDI deadline--it's hardly a paragon of sanity :stuck_out_tongue:

Reply:
https://github.com/flexflow/FlexFlow/issues/620